bindings/nodejs/src/structures/oic-client.cc

```diff
@@ -0,0 +1,128 @@
+}
+
+using namespace v8;
+
+static Local<Array> jsStringArrayFromStrSliceVector(
+    struct sol_vector *vector) {
+    Local<Array> jsArray = Nan::New<Array>(vector->len);
+    sol_str_slice *slice;
+    int index;
+    SOL_VECTOR_FOREACH_IDX(vector, slice, index) {
+        jsArray->Set(index,
+            Nan::New<String>(slice->data, slice->len).ToLocalChecked());
+    }
+    return jsArray;
+}
+
+Local<Object> SolOicClientResource::New(struct sol_oic_resource *resource) {
+    Local<Object> jsResource =
+        JSReffableHandle<SolOicClientResource>::New((void *)resource);
+
+    Nan::Set(jsResource, Nan::New("addr").ToLocalChecked(),
+        js_sol_network_link_addr(&(resource->addr)));
+    Nan::Set(jsResource, Nan::New("device_id").ToLocalChecked(),
+        js_DeviceIdFromSlice(&(resource->device_id)));
+    Nan::Set(jsResource, Nan::New("href").ToLocalChecked(),
+        Nan::New<String>(resource->href.data,
+            resource->href.len).ToLocalChecked());
+    Nan::Set(jsResource, Nan::New("interfaces").ToLocalChecked(),
+        jsStringArrayFromStrSliceVector(&(resource->interfaces)));
+    Nan::Set(jsResource, Nan::New("is_observing").ToLocalChecked(),
+        Nan::New(resource->is_observing));
+    Nan::Set(jsResource, Nan::New("observable").ToLocalChecked(),
+        Nan::New(resource->observable));
+    Nan::Set(jsResource, Nan::New("secure").ToLocalChecked(),
+        Nan::New(resource->secure));
+    Nan::Set(jsResource, Nan::New("types").ToLocalChecked(),
+        jsStringArrayFromStrSliceVector(&(resource->types)));
+
+    return jsResource;
+}
+
+const char *SolOicClient::jsClassName() {
+    return "SolOicClient";
+}
+
+void SolOicClient::ref(void *data) {}
+
+void SolOicClient::unref(void *data) {
+    sol_oic_client_del((sol_oic_client *)data);
+}
+
+OicCallbackData::OicCallbackData(): jsClient(0), callback(0),
+    hijackRefWasSuccessful(false) {}
+OicCallbackData::~OicCallbackData() {
+    if (jsClient) {
+        jsClient->Reset();
+        delete jsClient;
+    }
+    delete callback;
+    if (hijackRefWasSuccessful) {
+        hijack_unref();
+    }
+}
+
+bool OicCallbackData::init(Local<Object> _jsClient,
+    Local<Function> jsCallback) {
+    callback = new Nan::Callback(jsCallback);
+    if (!callback) {
+        Nan::ThrowError("OicCallbackData: Failed to allocate callback");
+        return false;
+    }
+    jsClient = new Nan::Persistent<Object>(_jsClient);
+    if (!jsClient) {
+        delete callback;
+        Nan::ThrowError("OicCallbackData: Failed to allocate client");
+        return false;
+    }
+
+    hijackRefWasSuccessful = hijack_ref();
+    return hijackRefWasSuccessful;
+}
+
+OicCallbackData *OicCallbackData::New(Local<Object> jsClient,
+    Local<Function> jsCallback) {
+    OicCallbackData *data = new OicCallbackData;
+    if (!data) {
+        Nan::ThrowError("Failed to allocate OicCallbackData");
+    } else if (!data->init(jsClient, jsCallback)) {
```

Comment:
I have a policy of throwing an exception at the point of failure - not later. Basically:
```
bool someOtherFunction() {
  if (uh-oh) {
    Nan::ThrowError("someOtherFunction: FAIL");
    return false;
  }
  return true;
}

bool someFunction() {
  if (uh-oh) {
    Nan::ThrowError("someFunction: FAIL");
    return false;
  }
  if (!someOtherFunction()) {
    return false;
  }
  return true;
}

NAN_METHOD(something) {
  if (uh-oh) {
    Nan::ThrowError("FAIL");
    return;
  }
  if (!someFunction()) {
    return;
  }
```
So, in this case, ```->init()``` will throw the exception and return ```false```. The ```false``` will propagate all the way back to the binding, which will simply return. This whole thing mimics the way JS tears down the stack when something throws an exception.